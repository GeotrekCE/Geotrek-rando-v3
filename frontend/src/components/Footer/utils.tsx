export const isLinkInternal = (urlString: string): boolean => {
  if (typeof window !== 'undefined') {
    const url = new URL(urlString);
    return url.hostname === window.location.hostname;
  }
  return false;
};

export const linkWithoutHost = (urlString: string): string => {
  const url = new URL(urlString);
  return url.pathname;
};
