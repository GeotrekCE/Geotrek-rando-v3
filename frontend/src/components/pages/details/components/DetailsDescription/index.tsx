export { DetailsDescription } from './DetailsDescription';
export { Step } from './DetailsDescription';
