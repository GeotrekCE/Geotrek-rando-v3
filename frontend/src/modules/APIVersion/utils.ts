import semver from 'semver';
import { APIVersion } from './interface';

export const isLowerOrEqualCurrentAPIVersion = (
  version: string,
  currentAPIVersion?: APIVersion,
) => {
  try {
    return semver.lte(version, currentAPIVersion?.version as string)
  } catch {
    return null;
  }
};
