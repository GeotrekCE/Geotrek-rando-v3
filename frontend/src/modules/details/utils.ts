import getNextConfig from 'next/config';

const {
  publicRuntimeConfig: {
    global: {
      displayObjectsRelatedToItinerantTreks: {
        POIs = true,
        touristicContents = true,
        sensitiveAreas = true,
        infrastructures = true,
        signages = true,
        services = true,
      } = {},
    },
  },
} = getNextConfig();

export const getObjectsRelatedToItinerantTreksToDisplay = (childrenIds: number[]) => {
  // The limitation concerns Itinerant treks
  if (childrenIds === null || childrenIds === undefined || childrenIds.length === 0) {
    return {
      displayRelatedPOIs: true,
      displayRelatedTouristicContents: true,
      displayRelatedSensitiveAreas: true,
      displayRelatedInfrastructures: true,
      displayRelatedSignages: true,
      displayRelatedServices: true,
    };
  }
  return {
    displayRelatedPOIs: POIs,
    displayRelatedTouristicContents: touristicContents,
    displayRelatedSensitiveAreas: sensitiveAreas,
    displayRelatedInfrastructures: infrastructures,
    displayRelatedSignages: signages,
    displayRelatedServices: services,
  };
};
