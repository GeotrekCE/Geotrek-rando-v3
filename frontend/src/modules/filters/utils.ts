import { uniqBy } from 'lodash';
import { getTouristicContentCategoryFilter } from 'modules/touristicContentCategory/connector';
import { getActivityFilter } from 'modules/activities/connector';
import { TouristicContentCategoryMapping } from 'modules/touristicContentCategory/interface';
import { getLabels } from 'modules/label/connector';
import { Label } from 'modules/label/interface';

import { getOutdoorPracticesFilter } from '../outdoorPractice/connector';
import { OutdoorPracticeChoices } from '../outdoorPractice/interface';
import { OutdoorRatingMapping } from '../outdoorRating/interface';
import { OutdoorRatingScale } from '../outdoorRatingScale/interface';
import { getTouristicEventTypesFilter } from '../touristicEventType/connector';
import { getAccessibilityFilter } from './accessibility/connector';
import { getCityFilter } from './city/connector';
import { getFiltersConfig } from './config';
import { getDifficultyFilter } from './difficulties/connector';
import { getCourseTypeFilter } from './courseType/connector';
import { getDistrictFilter } from './district/connector';
import {
  Filter,
  FilterConfigWithOptions,
  FilterState,
  FilterWithoutType,
  Option,
} from './interface';
import { getStructureFilter } from './structures/connector';
import { getThemeFilter } from './theme/connector';
import {
  ACCESSIBILITY_ID,
  CATEGORY_ID,
  CITY_ID,
  DISTRICT_ID,
  EVENT_ID,
  OUTDOOR_ID,
  PRACTICE_ID,
  ROUTE_ID,
  STRUCTURE_ID,
  THEME_ID,
} from './constant';

const adaptFilterConfigWithOptionsToFilter = (
  filterConfigWithOptions: FilterConfigWithOptions,
): Filter => ({
  id: filterConfigWithOptions.id,
  type: filterConfigWithOptions.type,
  options: filterConfigWithOptions.options.map(option => ({
    value: `${option.minValue}`,
    label: option.label,
  })),
});

const getFilterOptions = async (
  filterId: string,
  language: string,
): Promise<FilterWithoutType | null> => {
  switch (filterId) {
    case 'difficulty':
      return getDifficultyFilter(language);
    case PRACTICE_ID:
      return getActivityFilter(language);
    case CITY_ID:
      return getCityFilter(language);
    case DISTRICT_ID:
      return getDistrictFilter(language);
    case THEME_ID:
      return getThemeFilter(language);
    case ROUTE_ID:
      return getCourseTypeFilter(language);
    case ACCESSIBILITY_ID:
      return getAccessibilityFilter(language);
    case STRUCTURE_ID:
      return getStructureFilter(language);
    case CATEGORY_ID:
      return getTouristicContentCategoryFilter(language);
    case OUTDOOR_ID:
      return getOutdoorPracticesFilter(language);
    case EVENT_ID:
      return getTouristicEventTypesFilter(language);
    default:
      return null;
  }
};

const isElementNotNull = <ElementType>(element: ElementType | null): element is ElementType =>
  element !== null;

const getFilterAndAddType = async (
  filterId: string,
  filterType: 'SINGLE' | 'MULTIPLE',
  language: string,
): Promise<Filter | null> => {
  const filter = await getFilterOptions(filterId, language);
  if (filter === null) return null;
  return { ...filter, type: filterType };
};

const getFilters = async (language: string): Promise<Filter[]> => {
  const config = getFiltersConfig();
  const filters = await Promise.all(
    config.map(filterConfig => {
      if (filterConfig.options !== undefined) {
        return adaptFilterConfigWithOptionsToFilter(filterConfig);
      }
      return getFilterAndAddType(filterConfig.id, filterConfig.type, language);
    }),
  );
  return filters.filter(isElementNotNull);
};

const trekSpecificFilters = [
  'difficulty',
  'duration',
  'length',
  'ascent',
  ROUTE_ID,
  ACCESSIBILITY_ID,
  'labels',
];

export const commonFilters = [
  PRACTICE_ID,
  CATEGORY_ID,
  OUTDOOR_ID,
  EVENT_ID,
  THEME_ID,
  CITY_ID,
  DISTRICT_ID,
  STRUCTURE_ID,
];

export const getTreksFiltersState = (initialFiltersState: FilterState[]): FilterState[] =>
  initialFiltersState.filter(({ id }) => trekSpecificFilters.includes(id));

export const getFiltersState = async (language: string): Promise<FilterState[]> => {
  const filters = await getFilters(language);

  const labels = await getLabels(language);

  const labelsFilter = {
    id: 'labels',
    label: 'Autres',
    type: 'MULTIPLE',
    options: Object.values(labels)
      .filter(l => l.filter)
      .map((l: Label) => ({
        label: l.name,
        value: l.id.toString(),
        pictogramUrl: l.pictogramUri,
      })),
    selectedOptions: [],
  } as FilterState;

  return [
    ...filters.map(filter => ({
      ...filter,
      label: `search.filters.${filter.id}`,
      selectedOptions: [],
    })),
    labelsFilter,
  ];
};

const getTypesFiltersState = ({
  serviceId,
  touristicContentCategoryMapping,
}: {
  serviceId: string;
  touristicContentCategoryMapping: TouristicContentCategoryMapping;
}): FilterState[] => {
  const data = touristicContentCategoryMapping[parseInt(serviceId, 10)];

  if (!data) return [];

  return data.map(i => {
    return {
      id: `type-services-${i.id}`,
      category: i.category,
      label: i.label,
      type: 'MULTIPLE',
      options: i.values,
      selectedOptions: [],
    };
  });
};

const getOutdoorRatingFiltersState = ({
  practiceId,
  outdoorRatingMapping,
  outdoorRatingScale,
  outdoorPractice,
}: {
  practiceId: string;
  outdoorRatingMapping: OutdoorRatingMapping;
  outdoorRatingScale: OutdoorRatingScale[];
  outdoorPractice: OutdoorPracticeChoices;
}): FilterState[] => {
  const result: FilterState[] = [];
  const scales = outdoorRatingScale.filter(i => String(i.practice) === practiceId);

  if (scales.length > 0) {
    scales.forEach(scale => {
      const data = outdoorRatingMapping[scale.id];

      if (data)
        result.push({
          id: `type-outdoorRating-${String(scale.id)}`,
          category: String(outdoorPractice?.[scale.practice]?.name),
          label: scale?.name ?? 'Error',
          type: 'MULTIPLE',
          options: data.map(i => ({
            value: i.id,
            label: i.name,
          })),
          selectedOptions: [],
        });
    });
  }

  return result;
};

export const computeFiltersToDisplay = ({
  initialFiltersState,
  currentFiltersState,
  selectedFilterId,
  touristicContentCategoryMapping,
  outdoorRatingMapping,
  outdoorRatingScale,
  outdoorPractice,
}: {
  initialFiltersState: FilterState[];
  currentFiltersState: FilterState[];
  selectedFilterId: string;
  touristicContentCategoryMapping: TouristicContentCategoryMapping;
  outdoorRatingMapping: OutdoorRatingMapping;
  outdoorRatingScale: OutdoorRatingScale[];
  outdoorPractice: OutdoorPracticeChoices;
}): FilterState[] => {
  const trekPracticeFilter = currentFiltersState.find(i => i.id === PRACTICE_ID);
  const touristicContentFilter = currentFiltersState.find(i => i.id === CATEGORY_ID);
  const outdoorPracticeFilter = currentFiltersState.find(i => i.id === OUTDOOR_ID);

  const currentNumberOfPracticeOptionsSelected = trekPracticeFilter?.selectedOptions.length ?? 0;
  const currentNumberOfTouristicContentOptionsSelected =
    touristicContentFilter?.selectedOptions.length ?? 0;
  const currentNumberOfOutdoorPraticeOptionsSelected =
    outdoorPracticeFilter?.selectedOptions.length ?? 0;

  const filtersToAdd: FilterState[][] = [];

  // *** Calculate which filters to display ***
  // Treks filters
  if (currentNumberOfPracticeOptionsSelected > 0) {
    filtersToAdd.push(getTreksFiltersState(initialFiltersState));
  }
  // Services filters
  if (currentNumberOfTouristicContentOptionsSelected > 0 || selectedFilterId === CATEGORY_ID) {
    touristicContentFilter?.selectedOptions.forEach(selectedOptions => {
      filtersToAdd.push(
        getTypesFiltersState({
          serviceId: selectedOptions.value,
          touristicContentCategoryMapping,
        }),
      );
    });
  }
  // Outdoor filters
  if (currentNumberOfOutdoorPraticeOptionsSelected > 0 || selectedFilterId === OUTDOOR_ID) {
    outdoorPracticeFilter?.selectedOptions.forEach(selectedOptions => {
      filtersToAdd.push(
        getOutdoorRatingFiltersState({
          practiceId: selectedOptions.value,
          outdoorRatingMapping,
          outdoorRatingScale,
          outdoorPractice,
        }),
      );
    });
  }

  // Prevent old filters to display
  const filtersToAddIds = filtersToAdd.flat().map(i => i.id);
  const filtersToExcludeIds = currentFiltersState
    .filter(i => !filtersToAddIds.includes(i.id) && !commonFilters.includes(i.id))
    .map(i => i.id);
  const currentFiltersStateToDisplay = currentFiltersState.filter(
    i => !filtersToExcludeIds.includes(i.id),
  );

  return uniqBy([...currentFiltersStateToDisplay, ...filtersToAdd.flat()], 'id');
};

const getInitialFiltersStateWithRelevantFilters = ({
  initialFiltersState,
  initialOptions,
  touristicContentCategoryMapping,
  outdoorRatingMapping,
  outdoorRatingScale,
  outdoorPractice,
}: {
  initialFiltersState: FilterState[];
  initialOptions: { [filterId: string]: string[] | undefined };
  touristicContentCategoryMapping: TouristicContentCategoryMapping;
  outdoorRatingMapping: OutdoorRatingMapping;
  outdoorRatingScale: OutdoorRatingScale[];
  outdoorPractice: OutdoorPracticeChoices;
}): FilterState[] => {
  const initialStateWithOnlyCommon = initialFiltersState.filter(({ id }) =>
    commonFilters.includes(id),
  );
  const practices = initialOptions[PRACTICE_ID];
  const services = initialOptions[CATEGORY_ID];
  const outdoorPractices = initialOptions[OUTDOOR_ID];

  const result = [...initialStateWithOnlyCommon];

  if (Number(practices?.length) > 0) {
    result.push(...getTreksFiltersState(initialFiltersState));
  }

  if (Number(services?.length) > 0) {
    services?.forEach(service => {
      result.push(
        ...getTypesFiltersState({
          serviceId: service,
          touristicContentCategoryMapping,
        }),
      );
    });
  }

  if (Number(outdoorPractices?.length) > 0) {
    outdoorPractices?.forEach(outdoorPratice => {
      result.push(
        ...getOutdoorRatingFiltersState({
          practiceId: outdoorPratice,
          outdoorRatingMapping,
          outdoorRatingScale,
          outdoorPractice,
        }),
      );
    });
  }

  return result;
};

const sanitizeInitialOptions = (initialOptions: {
  [filterId: string]: string;
}): { [filterId: string]: string[] } =>
  Object.keys(initialOptions).reduce(
    (sanitizedOptions, key) => ({
      ...sanitizedOptions,
      ...(initialOptions[key] === '' ? {} : { [key]: initialOptions[key].split(',') }),
    }),
    {},
  );

export const getInitialFiltersStateWithSelectedOptions = ({
  initialFiltersState,
  initialOptions,
  touristicContentCategoryMapping,
  outdoorRatingMapping,
  outdoorRatingScale,
  outdoorPractice,
}: {
  initialFiltersState: FilterState[];
  initialOptions: { [filterId: string]: string };
  touristicContentCategoryMapping: TouristicContentCategoryMapping;
  outdoorRatingMapping: OutdoorRatingMapping;
  outdoorRatingScale: OutdoorRatingScale[];
  outdoorPractice: OutdoorPracticeChoices;
}): FilterState[] => {
  const sanitizedInitialOptions = sanitizeInitialOptions(initialOptions);
  const initialFiltersStateWithRelevantFilters = getInitialFiltersStateWithRelevantFilters({
    initialFiltersState,
    initialOptions: sanitizedInitialOptions,
    touristicContentCategoryMapping,
    outdoorRatingMapping,
    outdoorRatingScale,
    outdoorPractice,
  });
  return initialFiltersStateWithRelevantFilters.reduce<FilterState[]>(
    (initialStateWithSelectedOptions, currentFilterState) => {
      const selectedOptionsIds = sanitizedInitialOptions[currentFilterState.id];
      if (selectedOptionsIds === undefined) {
        return [...initialStateWithSelectedOptions, currentFilterState];
      }
      return [
        ...initialStateWithSelectedOptions,
        {
          ...currentFilterState,
          selectedOptions: currentFilterState.options.filter(({ value }) =>
            selectedOptionsIds.includes(`${value}`),
          ),
        },
      ];
    },
    [],
  );
};

const getTranslatedOptions = (selectedOptions: Option[], translatedOptions: Option[]) => {
  const translatedSelectedOptions = selectedOptions.map(selectedOption => {
    const translatedSelectedOption = translatedOptions.find(
      ({ value }) => value === selectedOption.value,
    );
    if (translatedSelectedOption === undefined) return null;
    return translatedSelectedOption;
  });
  return translatedSelectedOptions.filter(isElementNotNull);
};

export const getNewLanguageFiltersState = (
  filtersState: FilterState[],
  translatedInitialFiltersState: FilterState[],
): FilterState[] => {
  const translatedFiltersState = filtersState.map(filterState => {
    const translatedFilterState = translatedInitialFiltersState.find(
      ({ id }) => id === filterState.id,
    );
    if (translatedFilterState === undefined) return null;

    return {
      ...filterState,
      label: translatedFilterState.label,
      options: translatedFilterState.options,
      selectedOptions: getTranslatedOptions(
        filterState.selectedOptions,
        translatedFilterState.options,
      ),
    };
  });
  return translatedFiltersState.filter(isElementNotNull);
};

export const countFiltersSelected = (
  filtersState: FilterState[],
  filters: string[] | null = [],
  subFilters: string[] | string[][] | null = [],
): number => {
  const subFiltersToDisplay = filtersState.filter(({ id }) => subFilters?.flat().includes(id));
  const filtersToDisplay = filtersState.filter(
    ({ id }) => filters?.includes(id) ?? filters === null,
  );

  return [...subFiltersToDisplay, ...filtersToDisplay].reduce(
    (acc, state) => state.selectedOptions.length + acc,
    0,
  );
};
