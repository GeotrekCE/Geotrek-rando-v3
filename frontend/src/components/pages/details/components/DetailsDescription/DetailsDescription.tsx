import { FormattedMessage } from 'react-intl';
import styled, { css } from 'styled-components';
import { colorPalette, desktopOnly, getSpacing, shadow, typography } from 'stylesheet';
import parse from 'html-react-parser';
import { HtmlText } from '../../utils';

interface DetailsDescriptionProps {
  descriptionHtml: string;
  className?: string;
}

export const DetailsDescription: React.FC<DetailsDescriptionProps> = ({
  descriptionHtml,
  className,
}) => {
  return (
    <div
      id="details_description"
      className={`flex flex-col
      pt-6 desktop:pt-12
      border-solid border-greySoft border-b
      ${className ?? ''}`}
    >
      <p id="details_descriptionTitle" className="text-Mobile-H1 desktop:text-H2 font-bold">
        <FormattedMessage id="details.description" />
      </p>
      <div id="details_descriptionContent" className="mt-3 desktop:mt-4 mb-6 desktop:mb-12">
        <StyledListWithSteps>{parse(descriptionHtml)}</StyledListWithSteps>
      </div>
    </div>
  );
};

const StyledListWithSteps = styled(HtmlText)`
  & > ol {
    position: relative;
    list-style: none;
    counter-reset: item;
    margin: ${getSpacing(2)} 0;
    ${desktopOnly(css`
      margin: ${getSpacing(4)} 0;
    `)}
  }
  & > ol::before {
    content: ' ';
    background-color: ${colorPalette.primary1};
    width: 3px;
    position: absolute;
    top: 0;
    bottom: 0;
    left: 14px;
    z-index: -1;
    ${desktopOnly(css`
      left: 22px;
    `)}
  }
  & > ol > li {
    counter-increment: item;
    display: flex;
    align-items: center;
    padding-top: ${getSpacing(4)};
    ${desktopOnly(css`
      padding-top: ${getSpacing(10)};
    `)}
  }
  & > ol > li:first-child {
    padding: 0;
  }
  & > ol > li::before {
    content: counter(item);
    border-radius: 100%;
    width: ${getSpacing(8)};
    height: ${getSpacing(8)};
    flex: none;
    margin-right: ${getSpacing(3.5)};
    border: solid 3px ${colorPalette.primary1};
    font-weight: bold;
    color: ${colorPalette.primary1};
    background-color: white;
    display: flex;
    justify-content: center;
    align-items: center;
    box-shadow: ${shadow.medium};
    ${desktopOnly(css`
      width: ${getSpacing(12)};
      height: ${getSpacing(12)};
      margin-right: ${getSpacing(5.5)};
      ${typography.h4}
    `)}
  }
`;
