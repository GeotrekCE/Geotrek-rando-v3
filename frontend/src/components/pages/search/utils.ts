import { InfiniteData } from '@tanstack/react-query';
// @ts-ignore
import { remove as removeDiacritics } from 'diacritics';

import { routes } from 'services/routes';
import { FilterState, Option } from 'modules/filters/interface';
import { SearchResults, TrekResult } from 'modules/results/interface';
import { TouristicContentResult } from 'modules/touristicContent/interface';
import { MapResult } from 'modules/mapResults/interface';
import { OutdoorSiteResult } from '../../../modules/outdoorSite/interface';
import { TouristicEventResult } from '../../../modules/touristicEvent/interface';

/**
 * Interface of an objet representing a filter state informations
 * suitable for a backend query
 */
export interface QueryFilterState {
  /** Name of the API parameter */
  id: string;
  /** Selected values to query */
  selectedOptions: string[];
}

/** Transforms an option array into one suitable for the API query */
export const parseSelectedOptions = (selectedOptions: Option[]): string[] =>
  selectedOptions.map(selectedOption => selectedOption.value);

/** Transforms a FilterState object into one suitable for the API query */
export const parseFilter = (filter: FilterState): QueryFilterState => ({
  id: filter.id,
  selectedOptions: parseSelectedOptions(filter.selectedOptions),
});

/** Transforms a FilterState array into one suitable for the API query */
export const parseFilters = (filters: FilterState[]): QueryFilterState[] =>
  filters.map(parseFilter);

export const parseTextFilter = (textFilter: string | null): string | undefined =>
  textFilter !== null ? textFilter : undefined;

export const parseBboxFilter = (bboxFilter: string | null): string | undefined =>
  bboxFilter !== null ? bboxFilter : undefined;

/** Concatenates multiple SearchResults into one with all results concatenated */
export const concatResultsPages = (resultsPages: SearchResults[]): SearchResults | null => {
  if (resultsPages.length === 0) return null;

  const resultsNumber = resultsPages[0].resultsNumber;
  const previousPages = resultsPages[0].previousPages;
  const nextPages = resultsPages[0].nextPages;
  const results = resultsPages.reduce<
    (TrekResult | TouristicContentResult | OutdoorSiteResult | TouristicEventResult)[]
  >((resultsAcc, currentPage) => resultsAcc.concat(currentPage.results), []);

  return {
    resultsNumber,
    resultsNumberDetails: resultsPages[0].resultsNumberDetails,
    results,
    nextPages,
    previousPages,
  };
};

/** Formats the data of useInfiniteQuery to make it interpretable by the Search page */
export const formatInfiniteQuery = (
  infiniteQueryData: InfiniteData<SearchResults> | undefined,
): SearchResults | null => {
  if (infiniteQueryData === undefined) {
    return null;
  }
  return concatResultsPages(infiniteQueryData?.pages);
};

/** Generates the details page url related to a result */
export const generateResultDetailsUrl = (
  id: number | string,
  title: string,
  route: string = routes.TREK,
  parentId?: number,
): string => {
  const titleWithNoSpace = convertStringForSitemap(title);
  const detailsPageUrl = `${route}/${id}-${encodeURI(titleWithNoSpace)}${
    parentId ? `?parentId=${parentId}` : ''
  }`;

  return detailsPageUrl;
};

export const getHoverId = ({
  id,
  type,
}:
  | TrekResult
  | TouristicContentResult
  | MapResult
  | OutdoorSiteResult
  | TouristicEventResult): string => `SEARCH-${type}-${id}`;

export const convertStringForSitemap = (text: string): string =>
  removeDiacritics(
    (text || '')
      .replace(/ /g, '-')
      // eslint-disable-next-line
      .replace(/ /g, '-')
      .replace(/'/g, '-')
      .replace(/°/g, '')
      .replace(/«/g, '')
      .replace(/»/g, '')
      .replace(/"/g, '')
      .replace(/>/g, '')
      .replace(/</g, '')
      .replace(/’/g, '-')
      .replace(/–/g, '-')
      .replace(/%/g, '-')
      .replace(/\+/g, '')
      .replace(/\//g, '')
      .replace(/\\/g, ''),
  );
