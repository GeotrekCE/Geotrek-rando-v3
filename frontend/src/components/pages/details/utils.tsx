import { routes } from 'services/routes';
import styled, { css } from 'styled-components';
import { colorPalette, desktopOnly, getSpacing, typography } from 'stylesheet';
import { convertStringForSitemap, generateResultDetailsUrl } from '../search/utils';

export const HtmlText = styled.div`
  a {
    color: ${colorPalette.primary1};
    transition-property: all;
    transition-duration: 300ms;
    &:hover {
      color: ${colorPalette.primary1_light};
    }
  }
  h1 {
    ${typography.h2}
    ${desktopOnly(css`
      ${typography.h1}
    `)}
  }
  h2 {
    ${typography.h3}
    ${desktopOnly(css`
      ${typography.h2}
    `)}
  }
  h3 {
    ${typography.h4}
    ${desktopOnly(css`
      ${typography.h3}
    `)}
  }
  p {
    margin-bottom: ${getSpacing(2)};
    ${desktopOnly(css`
      margin-bottom: ${getSpacing(4)};
    `)}
  }
  h1,
  h2,
  h3,
  h4,
  h5 {
    margin: ${getSpacing(2)} 0;
    ${desktopOnly(css`
      margin: ${getSpacing(4)} 0;
    `)}
  }
  img {
    border-radius: ${getSpacing(4)};
    max-height: 75vh;
    width: auto;
    margin: ${getSpacing(1)} auto;
  }
  iframe {
    margin-top: ${getSpacing(5)};
    width: 100%;
    height: auto;
    ${desktopOnly(css`
      height: 70vh;
      margin-top: ${getSpacing(7)};
    `)}
  }

  b {
    font-weight: bold;
  }

  em {
    font-style: italic;
    ${typography.small}
    ${desktopOnly(css`
      ${typography.main}
    `)}
  }
  h6 {
    margin: 0 auto;
    text-align: center;
    margin-bottom: ${getSpacing(2)};
    ${desktopOnly(css`
      margin-bottom: ${getSpacing(4)};
    `)}
  }

  strong {
    font-weight: bold;
  }
  ul {
    margin-top: ${getSpacing(2)};
    margin-bottom: ${getSpacing(2)};
  }
  ul > li {
    padding-left: ${getSpacing(2)};
    display: list-item;
    list-style-type: disc;
    list-style-position: inside;
    ${desktopOnly(css`
      margin-bottom: ${getSpacing(1)};
    `)}
  }
`;

/** Computes an element top pixel and bottom pixel position */
export const getDimensions = (htmlElement: HTMLElement | null): { top: number; bottom: number } => {
  if (htmlElement === null) return { top: 0, bottom: 0 };
  const { height } = htmlElement.getBoundingClientRect();
  const elementTopPosition = htmlElement.offsetTop;
  const elementBottomPosition = elementTopPosition + height;

  return {
    top: elementTopPosition,
    bottom: elementBottomPosition,
  };
};

/** Generates the details page url for a trek part of an itinerance */
export const generateChildrenDetailsUrl = (
  id: string | number,
  title: string,
  parentId: string,
): string => {
  const detailsUrl = generateResultDetailsUrl(id, title);
  return `${detailsUrl}?parentId=${parentId}`;
};

export const generateTouristicContentUrl = (id: number | string, title: string): string => {
  const titleWithNoSpace = convertStringForSitemap(title);
  return `${routes.TOURISTIC_CONTENT}/${id}-${encodeURI(titleWithNoSpace)}`;
};

export const generateOutdoorSiteUrl = (id: number | string, title: string): string => {
  const titleWithNoSpace = convertStringForSitemap(title);
  return `${routes.OUTDOOR_SITE}/${id}-${encodeURI(titleWithNoSpace)}`;
};

export const generateOutdoorCourseUrl = (id: number | string, title: string): string => {
  const titleWithNoSpace = convertStringForSitemap(title);
  return `${routes.OUTDOOR_COURSE}/${id}-${encodeURI(titleWithNoSpace)}`;
};

export const generateTouristicEventUrl = (id: number | string, title: string): string => {
  const titleWithNoSpace = convertStringForSitemap(title);
  return `${routes.TOURISTIC_EVENT}/${id}-${encodeURI(titleWithNoSpace)}`;
};
