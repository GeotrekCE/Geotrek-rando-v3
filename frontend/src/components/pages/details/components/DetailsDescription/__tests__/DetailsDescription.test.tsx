import { render } from 'services/testing/reactTestingLibraryWrapper';
import { DetailsDescription } from '../DetailsDescription';
import { checkAndParseToList } from '../../../utils';
describe('DetailsDescription', () => {
  it('should display a well parsed description Element', () => {
    const [, intro, conclusion, steps] = checkAndParseToList(
      "<p>Du parking remonter la vall&eacute;e du S&eacute;l&eacute; par le sentier qui suit la rive gauche du torrent de Celse Ni&egrave;re. Apr&egrave;s une longue mont&eacute;e en faux plat, le sentier remonte des lacets raides jusqu'&agrave; la jonction avec le sentier du refuge du S&eacute;l&eacute; (1993 m).</p>\r\n<ol>\r\n<li>Prendre &agrave; droite le sentier du refuge du Pelvoux qui serpente au milieu des &eacute;boulis et de petites barres rocheuses, puis se poursuit en lacets dans des pentes d'herbe et de rochers peu inclin&eacute;es. Le refuge est invisible jusqu'&agrave; l'arriv&eacute;e mais l'approche est signal&eacute;e par des cairns (tas de pierres).</li>\r\n<li>Emprunter le m&ecirc;me itin&eacute;raire en sens inverse pour le retour.</li>\r\n</ol><br />Et ensuite vous &ecirc;tes arriv&eacute;s ! ;-)<br />Sympa ce petit tour, non ?",
    );
    const detailsDescription = render(
      <DetailsDescription intro={intro} steps={steps} conclusion={conclusion} />,
    );

    expect(detailsDescription).toMatchSnapshot();
  });
});
