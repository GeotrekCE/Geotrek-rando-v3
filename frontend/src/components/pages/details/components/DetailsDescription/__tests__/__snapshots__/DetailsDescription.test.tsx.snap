// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`DetailsDescription should display a well parsed description Element 1`] = `
Object {
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <div
        class="flex flex-col
      pt-6 desktop:pt-12
      border-solid border-greySoft border-b
      "
        id="details_description"
      >
        <p
          class="text-Mobile-H1 desktop:text-H2 font-bold"
          id="details_descriptionTitle"
        >
          Description
        </p>
        <div
          class="mt-3 desktop:mt-4 mb-6 desktop:mb-12"
          id="details_descriptionContent"
        >
          <div
            class="sc-bdfBQB sc-gsTEea hRTanl vdYaW"
          >
            <p>
              Du parking remonter la vallée du Sélé par le sentier qui suit la rive gauche du torrent de Celse Nière. Après une longue montée en faux plat, le sentier remonte des lacets raides jusqu'à la jonction avec le sentier du refuge du Sélé (1993 m).
            </p>
            

            <ol>
              

              <li>
                Prendre à droite le sentier du refuge du Pelvoux qui serpente au milieu des éboulis et de petites barres rocheuses, puis se poursuit en lacets dans des pentes d'herbe et de rochers peu inclinées. Le refuge est invisible jusqu'à l'arrivée mais l'approche est signalée par des cairns (tas de pierres).
              </li>
              

              <li>
                Emprunter le même itinéraire en sens inverse pour le retour.
              </li>
              

            </ol>
            <br />
            Et ensuite vous êtes arrivés ! ;-)
            <br />
            Sympa ce petit tour, non ?
          </div>
        </div>
      </div>
    </div>
  </body>,
  "container": <div>
    <div
      class="flex flex-col
      pt-6 desktop:pt-12
      border-solid border-greySoft border-b
      "
      id="details_description"
    >
      <p
        class="text-Mobile-H1 desktop:text-H2 font-bold"
        id="details_descriptionTitle"
      >
        Description
      </p>
      <div
        class="mt-3 desktop:mt-4 mb-6 desktop:mb-12"
        id="details_descriptionContent"
      >
        <div
          class="sc-bdfBQB sc-gsTEea hRTanl vdYaW"
        >
          <p>
            Du parking remonter la vallée du Sélé par le sentier qui suit la rive gauche du torrent de Celse Nière. Après une longue montée en faux plat, le sentier remonte des lacets raides jusqu'à la jonction avec le sentier du refuge du Sélé (1993 m).
          </p>
          

          <ol>
            

            <li>
              Prendre à droite le sentier du refuge du Pelvoux qui serpente au milieu des éboulis et de petites barres rocheuses, puis se poursuit en lacets dans des pentes d'herbe et de rochers peu inclinées. Le refuge est invisible jusqu'à l'arrivée mais l'approche est signalée par des cairns (tas de pierres).
            </li>
            

            <li>
              Emprunter le même itinéraire en sens inverse pour le retour.
            </li>
            

          </ol>
          <br />
          Et ensuite vous êtes arrivés ! ;-)
          <br />
          Sympa ce petit tour, non ?
        </div>
      </div>
    </div>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
